File: tests/python/test_config_defaults.py
#!/usr/bin/env python3
"""契约测试：AppConfig JSON 结构校验。匹配 Go types.AppConfig。"""
import json
import os
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent


def _config_path():
    """解析 AppConfig 实际落点（与 Go 端 configPath() 对齐）。

    Wails 3 迁移（ADR-0001 §7 #5）后配置落在 os.UserConfigDir()/YSM-Model-Manager/ysm_config.json，
    仓库根的 ysm_config.json 已被 .gitignore 排除、不再作为 canonical 位置。
    返回首个存在的候选路径；都不存在（首次运行/纯净环境）返回 None。
    """
    if sys.platform == "win32":
        base = os.environ.get("APPDATA")
    elif sys.platform == "darwin":
        base = os.path.expanduser("~/Library/Application Support")
    else:
        base = os.path.expanduser("~/.config")
    candidates = []
    if base:
        candidates.append(Path(base) / "YSM-Model-Manager" / "ysm_config.json")
    candidates.append(ROOT / "ysm_config.json")  # 遗留位置（迁移兼容）
    for p in candidates:
        if p.exists():
            return p
    return None

# 对应 Go types.AppConfig 的 json tag
STRING_FIELDS = [
    "filesRoot", "ysmRoot", "repoRoot", "resourcepackRoot", "shaderpackRoot",
    "schematicRoot", "litematicRoot", "mmdRoot", "vrcRoot", "mcRoot",
    "linkMode", "theme", "mirror",
]
INT_FIELDS = [
    "winX", "winY", "winW", "winH", "winRelX", "winRelY", "winScrW", "winScrH",
]
ALWAYS_REQUIRED = [
    "filesRoot", "ysmRoot", "resourcepackRoot", "shaderpackRoot",
    "schematicRoot", "mmdRoot", "vrcRoot", "mcRoot",
    "linkMode", "theme", "mirror",
    "winX", "winY", "winW", "winH", "winRelX", "winRelY", "winScrW", "winScrH",
]
STRING_FIELDS = ALWAYS_REQUIRED[:11] + ["litematicRoot", "repoRoot"]
INT_FIELDS = ALWAYS_REQUIRED[11:] + ["voxelMaxBlocks"]


def validate():
    errors = []
    cfg = _config_path()
    if cfg is None:
        # 首次运行或纯净环境无配置文件，属合法状态，不视为违规
        return errors
    data = json.loads(cfg.read_text("utf-8"))

    for field in ALWAYS_REQUIRED:
        if field not in data:
            errors.append(f"MISSING: '{field}' not in config")
            continue

    for field in STRING_FIELDS:
        val = data.get(field)
        if val is not None and not isinstance(val, str):
            errors.append(f"TYPE: '{field}' must be string (got {type(val).__name__})")

    for field in INT_FIELDS:
        val = data.get(field)
        if val is not None and not isinstance(val, int):
            errors.append(f"TYPE: '{field}' must be int (got {type(val).__name__})")

    vm = data.get("voxelMaxBlocks")
    if vm is not None and not isinstance(vm, int):
        errors.append(f"TYPE: 'voxelMaxBlocks' must be int (got {type(vm).__name__})")

    link_mode = data.get("linkMode", "")
    if link_mode and link_mode not in ("copy", "hardlink", "symlink", ""):
        errors.append(f"VALUE: 'linkMode' must be copy/hardlink/symlink (got '{link_mode}')")

    theme = data.get("theme", "")
    valid_themes = {"cyber", "warm", "pro", "default-dark", "mint", "ocean", ""}
    if theme and theme not in valid_themes:
        errors.append(f"VALUE: 'theme' must be one of {valid_themes} (got '{theme}')")

    vr = data.get("vrcRoot", "")
    mm = data.get("mmdRoot", "")
    if vr and not mm:
        pass  # VRChat fallback to MMD is valid

    return errors


def main():
    errors = validate()
    if errors:
        sys.stdout.buffer.write(f"FAILED: {len(errors)} violation(s)\n\n".encode("utf-8"))
        for e in errors:
            sys.stdout.buffer.write(f"  {e}\n".encode("utf-8"))
        sys.exit(1)
    else:
        sys.stdout.buffer.write(b"OK: config schema checks passed\n")


if __name__ == "__main__":
    main()

File: tests/test_config_defaults.mjs
#!/usr/bin/env node
/**
 * 契约测试：AppConfig JSON 结构校验。匹配 Go types.AppConfig。
 * 由 tests/python/test_config_defaults.py 迁移（2026-08-03），校验逻辑逐点保真。
 */
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

/**
 * 解析 AppConfig 实际落点（与 Go 端 configPath() 对齐）。
 * Wails 3 迁移（ADR-001 §7 #5）后配置落在 os.UserConfigDir()/YSM-Model-Manager/ysm_config.json，
 * 仓库根的 ysm_config.json 已被 .gitignore 排除、不再作为 canonical 位置。
 * 返回首个存在的候选路径；都不存在（首次运行/纯净环境）返回 null。
 */
function configPath() {
  let base = null;
  if (process.platform === 'win32') {
    base = process.env.APPDATA ?? null;
  } else if (process.platform === 'darwin') {
    base = path.join(os.homedir(), 'Library', 'Application Support');
  } else {
    base = path.join(os.homedir(), '.config');
  }
  const candidates = [];
  if (base) candidates.push(path.join(base, 'YSM-Model-Manager', 'ysm_config.json'));
  candidates.push(path.join(ROOT, 'ysm_config.json')); // 遗留位置（迁移兼容）
  for (const p of candidates) {
    if (fs.existsSync(p)) return p;
  }
  return null;
}

// 对应 Go types.AppConfig 的 json tag
const ALWAYS_REQUIRED = [
  'filesRoot', 'ysmRoot', 'resourcepackRoot', 'shaderpackRoot',
  'schematicRoot', 'mmdRoot', 'vrcRoot', 'mcRoot',
  'linkMode', 'theme', 'mirror',
  'winX', 'winY', 'winW', 'winH', 'winRelX', 'winRelY', 'winScrW', 'winScrH',
];
const STRING_FIELDS = ALWAYS_REQUIRED.slice(0, 11).concat(['litematicRoot', 'repoRoot']);
const INT_FIELDS = ALWAYS_REQUIRED.slice(11).concat(['voxelMaxBlocks']);

function validate() {
  const errors = [];
  const cfg = configPath();
  if (cfg === null) {
    // 首次运行或纯净环境无配置文件，属合法状态，不视为违规
    return errors;
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(cfg, 'utf-8'));
  } catch (e) {
    errors.push(`SYNTAX: ${cfg} 解析失败: ${e.message}`);
    return errors;
  }

  for (const field of ALWAYS_REQUIRED) {
    if (!(field in data)) {
      errors.push(`MISSING: '${field}' not in config`);
    }
  }

  for (const field of STRING_FIELDS) {
    const val = data[field];
    if (val !== undefined && val !== null && typeof val !== 'string') {
      errors.push(`TYPE: '${field}' must be string (got ${typeof val})`);
    }
  }

  for (const field of INT_FIELDS) {
    const val = data[field];
    if (val !== undefined && val !== null && typeof val !== 'number') {
      errors.push(`TYPE: '${field}' must be int (got ${typeof val})`);
    }
  }

  const vm = data['voxelMaxBlocks'];
  if (vm !== undefined && vm !== null && typeof vm !== 'number') {
    errors.push(`TYPE: 'voxelMaxBlocks' must be int (got ${typeof vm})`);
  }

  const link_mode = data['linkMode'] ?? '';
  if (link_mode && !['copy', 'hardlink', 'symlink', ''].includes(link_mode)) {
    errors.push(`VALUE: 'linkMode' must be copy/hardlink/symlink (got '${link_mode}')`);
  }

  const theme = data['theme'] ?? '';
  const valid_themes = new Set(['cyber', 'warm', 'pro', 'default-dark', 'mint', 'ocean', '']);
  if (theme && !valid_themes.has(theme)) {
    errors.push(`VALUE: 'theme' must be one of ${[...valid_themes]} (got '${theme}')`);
  }

  return errors;
}

const errors = validate();
if (errors.length) {
  console.error(`FAILED: ${errors.length} violation(s)\n`);
  for (const e of errors) console.error(`  ${e}`);
  process.exit(1);
} else {
  console.log('OK: config schema checks passed');
}
